### PR TITLE
Dpp 1374 multi learner specs

### PR DESCRIPTION
**Pull request by Pauljgraham**
Added multi learner transfer specs

### CODE REVIEW

File: src/AcceptanceTesting/SpecByExample/SFA.DAS.Payments.AcceptanceTests/Assertions/PaymentsAndEarningsRules/EmployersLevyAccountDebitedForUlnViaTransferRule.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
+﻿using System;
```
Comment (davehowlett) [inline src/AcceptanceTesting/SpecByExample/SFA.DAS.Payments.AcceptanceTests/Assertions/PaymentsAndEarningsRules/EmployersLevyAccountDebitedForUlnViaTransferRule.cs]: unused using - can be removed

File: src/AcceptanceTesting/SpecByExample/SFA.DAS.Payments.AcceptanceTests/Assertions/PaymentsAndEarningsRules/EmployersLevyAccountDebitedForUlnViaTransferRule.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+using System.Linq;
+using SFA.DAS.Payments.AcceptanceTests.Contexts;
+using SFA.DAS.Payments.AcceptanceTests.ExecutionManagers;
```
Comment (davehowlett) [inline src/AcceptanceTesting/SpecByExample/SFA.DAS.Payments.AcceptanceTests/Assertions/PaymentsAndEarningsRules/EmployersLevyAccountDebitedForUlnViaTransferRule.cs]: unused using - can be removed

File: src/AcceptanceTesting/SpecByExample/SFA.DAS.Payments.AcceptanceTests/Assertions/PaymentsAndEarningsRules/EmployersLevyAccountDebitedForUlnViaTransferRule.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+                    .ToArray();
+
+                AssertResultsForPeriod(prevPeriod, allPayments);
+            }
+        }
+
+        protected override string FormatAssertionFailureMessage(PeriodValue period, decimal actualPaymentInPeriod)
+        {
+            var employerUlnPeriod = (EmployerAccountUlnPeriodValue) period;
+            var specPeriod = period.PeriodName.ToPeriodDateTime().AddMonths(1).ToPeriodName();
```
Comment (davehowlett) [inline src/AcceptanceTesting/SpecByExample/SFA.DAS.Payments.AcceptanceTests/Assertions/PaymentsAndEarningsRules/EmployersLevyAccountDebitedForUlnViaTransferRule.cs]: specPeriod isn't used and can be deleted.

File: src/AcceptanceTesting/SpecByExample/SFA.DAS.Payments.AcceptanceTests/DataCollectors/EarningsCollector.cs [hunk 4]
```diff
@@ -38,7 +38,8 @@ private static PeriodisedValuesEntity[] ReadEarningsFromDeds()
             {
                 var command = "SELECT "
```
Comment (davehowlett) [inline src/AcceptanceTesting/SpecByExample/SFA.DAS.Payments.AcceptanceTests/DataCollectors/EarningsCollector.cs]: not a fan of concatenated strings. Could this be re-written as a verbatim string to make it more readable

Comment (davehowlett) [reply src/AcceptanceTesting/SpecByExample/SFA.DAS.Payments.AcceptanceTests/DataCollectors/EarningsCollector.cs]: command could also be a constant

File: src/AcceptanceTesting/SpecByExample/SFA.DAS.Payments.AcceptanceTests/DataCollectors/SubmissionDataLockResultCollector.cs [hunk 5]
```diff
@@ -31,6 +31,7 @@ private static SubmissionDataLockResultEntity[] ReadDataLockResultsFromDeds()
                 var query = "SELECT "
```
Comment (davehowlett) [inline src/AcceptanceTesting/SpecByExample/SFA.DAS.Payments.AcceptanceTests/DataCollectors/SubmissionDataLockResultCollector.cs]: can be a verbatim string and a constant

File: src/AcceptanceTesting/SpecByExample/SFA.DAS.Payments.AcceptanceTests/TableParsers/CommitmentsTableParser.cs [hunk 6]
```diff
@@ -194,6 +197,12 @@ private static CommitmentReferenceData ParseCommitmentsTableRow(TableRow row, Co
                 throw new ArgumentException($"'{row[structure.EffectiveToIndex]}' is not a valid effective to");
             }
 
+            DateTime? transferApprovalDate = null;
+            if (structure.TransferApprovalDateIndex > -1 && !TryParseNullableDateTime(row[structure.TransferApprovalDateIndex], out transferApprovalDate))
+            {
+                throw new ArgumentException($"'{row[structure.TransferApprovalDateIndex]}' is not a valid effective from");
+            }
+
             var standardCode = row.ReadRowColumnValue<long>(structure.StandardCodeIndex, "standard code", Defaults.StandardCode);
```
Comment (davehowlett) [inline src/AcceptanceTesting/SpecByExample/SFA.DAS.Payments.AcceptanceTests/TableParsers/CommitmentsTableParser.cs:206]: generic <long> not needed here.

**Review by davehowlett — CHANGES_REQUESTED** [review]
couple of minor bits of formatting etc but all the tests pass and looks ok otherwise

File: src/AcceptanceTesting/SpecByExample/SFA.DAS.Payments.AcceptanceTests/Assertions/PaymentsAndEarningsRules/EmployersLevyAccountDebitedForUlnViaTransferRule.cs (hunk 3 above)
Comment (Pauljgraham) [reply src/AcceptanceTesting/SpecByExample/SFA.DAS.Payments.AcceptanceTests/Assertions/PaymentsAndEarningsRules/EmployersLevyAccountDebitedForUlnViaTransferRule.cs]: actually specPeriod should have been used in the assertion message

File: src/AcceptanceTesting/SpecByExample/SFA.DAS.Payments.AcceptanceTests/DataCollectors/EarningsCollector.cs (hunk 4 above)
Comment (Pauljgraham) [reply src/AcceptanceTesting/SpecByExample/SFA.DAS.Payments.AcceptanceTests/DataCollectors/EarningsCollector.cs]: Made the command const. but refrained from making it a verbetim string as it doesn't add any value (in this case)

File: src/AcceptanceTesting/SpecByExample/SFA.DAS.Payments.AcceptanceTests/DataCollectors/SubmissionDataLockResultCollector.cs (hunk 5 above)
Comment (Pauljgraham) [reply src/AcceptanceTesting/SpecByExample/SFA.DAS.Payments.AcceptanceTests/DataCollectors/SubmissionDataLockResultCollector.cs]: made into a const

File: src/AcceptanceTesting/SpecByExample/SFA.DAS.Payments.AcceptanceTests/TableParsers/CommitmentsTableParser.cs (hunk 6 above)
Comment (Pauljgraham) [reply src/AcceptanceTesting/SpecByExample/SFA.DAS.Payments.AcceptanceTests/TableParsers/CommitmentsTableParser.cs:206]: I know resharper says it's not needed, but it's unclear what type is returned unless you look at the implementation of the class. Would rather keep it here

**Review by davehowlett — APPROVED** [review]
changes look fine

File: src/AcceptanceTesting/SpecByExample/SFA.DAS.Payments.AcceptanceTests/TableParsers/CommitmentsTableParser.cs (hunk 6 above)
Comment (carlbm) [reply src/AcceptanceTesting/SpecByExample/SFA.DAS.Payments.AcceptanceTests/TableParsers/CommitmentsTableParser.cs:206]: I don't see a generic in the changed code?

**Review by carlbm — APPROVED** [review]
I haven't run the tests, but assuming that no more fail and that the 3 new tests pass it's approved

**Review by davehowlett — APPROVED** [review]
changes look ok now.

**Comment by satya-sanikommu** [comment]
Review and Testing completed. Ready to merge

**Pull request merged**